### PR TITLE
Update CancelToken.js

### PR DESCRIPTION
**Pull request by njlr**
In order to make this library work correctly, the user must pass the cancellation token around and check its state. To make this more convenient, I would propose a method to "check" the token. 

Usage like this: 

```
const sleep = ms => new Promise((resolve) => {
  setTimeout(() => {
    resolve(ms);
  }, ms);
});

const token = new CancelToken();

// This token has a 3 second time-out
sleep(3000).then(token.cancel());

const p = sleep(5000).then(() => {
  // Check the token has not been cancelled; 
  // without this, "Hello, world. " is always printed. 
  token.check();
  console.log('Hello, world. ');
});

const q = cancellablePromise(p, token);

q.then(console.log).catch(console.error);
```

### CODE REVIEW

**Comment by cyrilchapon** [comment]
Hey @njlr, thanks for the PR.

> the user must pass the cancellation token around and check its state

I think there is a misunderstanding here. The point of this lib is to wrap a Promise inside another, and actually make the other (outside one) cancellable (I.E aforementionned "non-instrusive" strategy). The outer Promise will resolve (or reject) if underlying promise resolves (or rejects) before the token was cancelled. The outer Promise will reject with `CancelError` if token is cancelled before underlying Promise is resolved (or rejected)

The correct way to implement what you are trying is simply like so :

```
const sleep = ms => new Promise((resolve) => {
  setTimeout(() => {
    resolve(ms);
  }, ms);
});

const token = new CancelToken();

// README : I fixed a typo here, you were cancelling token directly synchronously
sleep(3000).then(() => token.cancel());

const p = sleep(5000);

const q = cancellablePromise(p, token).then(() => {
  console.log('Hello, world. ');
});
```

In other words, do attach `.then` and `.catch` to `cancellablePromise` return, not the underlying promise.

**Comment by njlr** [comment]
Perhaps I chose a poor example... imagine that `p` was something like: 

```
async () => {
  await expensiveTask();
  await expensiveTask();
  await expensiveTask();
}
```

Then you would only get good cancellation behaviour doing something like this: 

```
async () => {
  token.check();
  await expensiveTask();
  token.check();
  await expensiveTask();
  token.check();
  await expensiveTask();
}
```

I suppose you could wrap each internal `expensiveTask` call: 

```
async () => {
  await cancellablePromise(expensiveTask(), token);
  await cancellablePromise(expensiveTask(), token);
  await cancellablePromise(expensiveTask(), token);
}
```

Is that the intended use of this library? 

This approach would not work for slow synchronous code: 

```
async () => {
  for (let i = 0; i < 10000; i++) {
    token.check();
    expensiveSyncTask();
  }
}
```

**Comment by njlr** [comment]
Actually, this won't work: 

```
async () => {
  for (let i = 0; i < 10000; i++) {
    token.check();
    expensiveSyncTask();
  }
}
```

This is because the event loop is blocked. Instead, it should look like this: 

```
async () => {
  for (let i = 0; i < 10000; i++) {
    await token.check();
    expensiveSyncTask();
  }
}
```

And `check` should look like this: 

```
check () {
  return Promise.race([ this.promise, Promise.resolve() ]); 
}
```

**Comment by cyrilchapon** [comment]
Thanks for clarifying 😄 

As far as I understand, you're saying that
```
const p async () => {
  await expensiveTask();
  await expensiveTask();
  await expensiveTask();
}

const ct = new CancelToken()

timeout(() => ct.cancel(), 3000)

const q = cancellablePromise(p, ct)

```
... wont work, because the event loop is not being released between (or even **during**) each tasks ?
Are you facing a situation like this ? Maybe you can elaborate on an actual real use-case ?

----------------

> Is that the intended use of this library?
> ```
> async () => {
>   await cancellablePromise(expensiveTask(), token);
>   await cancellablePromise(expensiveTask(), token);
>   await cancellablePromise(expensiveTask(), token);
> }
> ```

No, actually your sequential `await`s can be seen like sequential `then`s, and I'm personaly using this in a looooooong list of promises.

-------------

I'm even actualy using the same cancel token in 2 parallels looong promise.then().then().then() list.

Though, I wonder if

```
const p async () => {
  await expensiveTask();
  await expensiveTask();
  await expensiveTask();
}
```

is understood by v8 like

```
const p = () => {
  return expensiveTask()
    .then(expensiveTask)
    .then(expensiveTask)
}
```

Or more like

```
const p = () => {
  return expensiveTask()
    .then(() => {
      return expensiveTask()
        .then(() => {
          return expensiveTask()
        })
    })
}
```

-------

I'm interested in improving this library, though, 2 things :
- I'm only using it to wrap http calls, or file i/o Promises. I thought that the whole point of async code in Node.js was exactly `releasing the event loop` while doing other things (like, eg. running a timeout that will resolve a promise that will race against the first). Do you have any example of `expensive task as-a-promise that doesn't release the event loop` ?
- If we improve that, this **must** IMHO be transparent for the user. The `check()` stuff has to be internal, and the API I was designing *just rejects the wrapping Promise when you `cancel()` the token*. Not less, not more. Do you think it's possible with your idea ? (actualy, the `race` stuff is precisely what is implemented inside cancellablePromise... The `cancel` rejects a dumb Promise, and `CancellablePromise` just race that one with the underlying actual Promise...

**Comment by cyrilchapon** [comment]
I should add that, in my first idea, I was not going to **actually cancel** the process or the first Promise I did wrap, but more like just ignore it's result.
